File: backend/Services/CvParsingService.cs
using System.Text.RegularExpressions;
using System.Text.Json;
using DocumentFormat.OpenXml.Packaging;
using UglyToad.PdfPig;
using CareerMail.Api.Data;
using CareerMail.Api.Models.Entities;
using Microsoft.EntityFrameworkCore;

namespace CareerMail.Api.Services;

public interface ICvParsingService
{
    Task<CvProfile> ParseAndSaveCvAsync(long userId, Stream fileStream, string fileName);
    CvProfile ExtractProfileFromText(string rawText, long userId, string fileName);
}

public class CvParsingService : ICvParsingService
{
    private readonly IGeminiCvService _geminiCvService;
    private readonly AppDbContext _context;
    private readonly IConfiguration _configuration;
    private readonly ILogger<CvParsingService> _logger;

    private static readonly string[] SkillTaxonomy = new[]
    {
        "C#", ".NET", "ASP.NET", "EF Core", "Entity Framework",
        "React", "TypeScript", "JavaScript", "HTML", "CSS", "Tailwind CSS", "Next.js", "Vue", "Angular", "Sass",
        "Python", "Django", "FastAPI", "Flask", "PyTorch", "TensorFlow", "Pandas", "NumPy", "Scikit-Learn",
        "Java", "Spring Boot", "Kotlin", "Swift", "Flutter", "React Native",
        "C++", "Rust", "Go", "Golang", "PHP", "Laravel", "Ruby", "Ruby on Rails",
        "SQL", "PostgreSQL", "MySQL", "SQLite", "MongoDB", "Redis", "Elasticsearch", "Cassandra",
        "Docker", "Kubernetes", "AWS", "Azure", "GCP", "Cloud", "DevOps", "Terraform", "CI/CD", "GitHub Actions",
        "REST API", "GraphQL", "gRPC", "Microservices", "System Design", "WebSockets", "OAuth", "JWT",
        "Git", "Linux", "Unix", "Bash", "Shell", "Jira", "Agile", "Scrum",
        "Unit Testing", "Integration Testing", "Jest", "Cypress", "Selenium",
        "Machine Learning", "Artificial Intelligence", "Deep Learning", "NLP", "Data Science", "Data Engineering",
        "UI/UX", "Figma", "Cybersecurity", "Embedded Systems", "Product Management"
    };

    private static readonly string[] RoleTaxonomy = new[]
    {
        "Software Engineer", "Full Stack Developer", "Frontend Developer", "Backend Developer",
        "Web Developer", "Software Developer", "Dotnet Developer", "React Developer", "Python Developer",
        "Data Scientist", "Data Analyst", "Data Engineer", "Machine Learning Engineer", "AI Engineer",
        "DevOps Engineer", "Cloud Engineer", "Systems Architect", "Solutions Architect",
        "Product Manager", "Project Manager", "Scrum Master", "Technical Lead", "Engineering Manager",
        "QA Engineer", "Mobile Developer", "iOS Developer", "Android Developer", "UI/UX Designer"
    };

    public CvParsingService(
        IGeminiCvService geminiCvService,
        AppDbContext context,
        IConfiguration configuration,
        ILogger<CvParsingService> logger)
    {
        _geminiCvService = geminiCvService;
        _context = context;
        _configuration = configuration;
        _logger = logger;
    }

    public async Task<CvProfile> ParseAndSaveCvAsync(long userId, Stream fileStream, string fileName)
    {
        string rawText = string.Empty;
        var ext = Path.GetExtension(fileName).ToLowerInvariant();

        try
        {
            if (ext == ".pdf")
            {
                using var pdf = PdfDocument.Open(fileStream);
                var pageTexts = new List<string>();
                foreach (var page in pdf.GetPages())
                {
                    pageTexts.Add(page.Text);
                }
                rawText = string.Join("\n", pageTexts);
            }
            else if (ext == ".docx")
            {
                using var wordDoc = WordprocessingDocument.Open(fileStream, false);
                rawText = wordDoc.MainDocumentPart?.Document?.Body?.InnerText ?? string.Empty;
            }
            else
            {
                using var reader = new StreamReader(fileStream);
                rawText = await reader.ReadToEndAsync();
            }
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "Failed to parse binary file structure for {FileName}", fileName);
            rawText = $"Failed to parse binary file structure: {ex.Message}. (Fallback text extraction applied)";
        }

        if (string.IsNullOrWhiteSpace(rawText))
        {
            rawText = $"Extracted text from {fileName}";
        }

        // Check if Gemini API key is available (User DB Key or Configuration)
        var user = await _context.Users.AsNoTracking().FirstOrDefaultAsync(u => u.Id == userId);
        var geminiKey = user?.GeminiApiKey ?? _configuration["Gemini:ApiKey"] ?? Environment.GetEnvironmentVariable("GEMINI_API_KEY");

        if (!string.IsNullOrWhiteSpace(geminiKey))
        {
            _logger.LogInformation("Invoking Gemini AI for CV skill extraction & normalization for user {UserId}", userId);
            var geminiResult = await _geminiCvService.ExtractStructuredCvAsync(rawText, geminiKey);
            if (geminiResult != null)
            {
                return BuildProfileFromGemini(geminiResult, rawText, userId, fileName);
            }
            _logger.LogWarning("Gemini analysis returned null, falling back to deterministic taxonomy extraction.");
        }

        return ExtractProfileFromText(rawText, userId, fileName);
    }

    private CvProfile BuildProfileFromGemini(GeminiStructuredCvProfileDto gemini, string rawText, long userId, string fileName)
    {
        var allSkills = new HashSet<string>(StringComparer.OrdinalIgnoreCase);

        if (gemini.NormalisedSkills != null) foreach (var s in gemini.NormalisedSkills) if (!string.IsNullOrWhiteSpace(s)) allSkills.Add(s.Trim());
        if (gemini.TechnicalSkills != null) foreach (var s in gemini.TechnicalSkills) if (!string.IsNullOrWhiteSpace(s)) allSkills.Add(s.Trim());
        if (gemini.ProgrammingLanguages != null) foreach (var s in gemini.ProgrammingLanguages) if (!string.IsNullOrWhiteSpace(s)) allSkills.Add(s.Trim());
        if (gemini.Frameworks != null) foreach (var s in gemini.Frameworks) if (!string.IsNullOrWhiteSpace(s)) allSkills.Add(s.Trim());
        if (gemini.Databases != null) foreach (var s in gemini.Databases) if (!string.IsNullOrWhiteSpace(s)) allSkills.Add(s.Trim());
        if (gemini.CloudPlatforms != null) foreach (var s in gemini.CloudPlatforms) if (!string.IsNullOrWhiteSpace(s)) allSkills.Add(s.Trim());
        if (gemini.Tools != null) foreach (var s in gemini.Tools) if (!string.IsNullOrWhiteSpace(s)) allSkills.Add(s.Trim());
        if (gemini.DataSkills != null) foreach (var s in gemini.DataSkills) if (!string.IsNullOrWhiteSpace(s)) allSkills.Add(s.Trim());
        if (gemini.AiMlSkills != null) foreach (var s in gemini.AiMlSkills) if (!string.IsNullOrWhiteSpace(s)) allSkills.Add(s.Trim());

        var targetRoles = (gemini.JobTitles != null && gemini.JobTitles.Count > 0)
            ? gemini.JobTitles.Where(r => !string.IsNullOrWhiteSpace(r)).Select(r => r.Trim()).Distinct().ToList()
            : new List<string> { "Software Engineer" };

        string education = (gemini.Education != null && gemini.Education.Count > 0)
            ? string.Join(", ", gemini.Education.Take(2))
            : "Bachelor's Degree";

        bool isRemotePreferred = Regex.IsMatch(rawText, @"\b(Remote|Work from home|Hybrid)\b", RegexOptions.IgnoreCase);
        string preferredLocation = "Flexible / Remote";
        var locMatch = Regex.Match(rawText, @"\b(London|Manchester|New York|San Francisco|Austin|Berlin|Toronto|Remote|UK|United Kingdom|USA)\b", RegexOptions.IgnoreCase);
        if (locMatch.Success)
        {
            preferredLocation = locMatch.Value;
        }

        return new CvProfile
        {
            UserId = userId,
            FileName = fileName,
            RawText = rawText,
            ExtractedSkillsJson = JsonSerializer.Serialize(allSkills.ToList()),
            TargetRolesJson = JsonSerializer.Serialize(targetRoles),
            ExperienceYears = Math.Max(0, gemini.ExperienceYears ?? 0),
            EducationLevel = education,
            PreferredLocation = preferredLocation,
            IsRemotePreferred = isRemotePreferred,
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow
        };
    }

    public CvProfile ExtractProfileFromText(string rawText, long userId, string fileName)
    {
        var extractedSkills = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        foreach (var skill in SkillTaxonomy)
        {
            string pattern;
            if (skill.Equals(".NET", StringComparison.OrdinalIgnoreCase) || skill.Equals("C#", StringComparison.OrdinalIgnoreCase) || skill.Equals("C++", StringComparison.OrdinalIgnoreCase))
            {
                pattern = Regex.Escape(skill);
            }
            else
            {
                pattern = $@"\b{Regex.Escape(skill)}\b";
            }

            if (Regex.IsMatch(rawText, pattern, RegexOptions.IgnoreCase))
            {
                extractedSkills.Add(skill);
            }
        }

        var matchedRoles = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        foreach (var role in RoleTaxonomy)
        {
            if (Regex.IsMatch(rawText, $@"\b{Regex.Escape(role)}\b", RegexOptions.IgnoreCase))
            {
                matchedRoles.Add(role);
            }
        }

        if (matchedRoles.Count == 0)
        {
            if (extractedSkills.Contains("C#") || extractedSkills.Contains(".NET"))
            {
                matchedRoles.Add("Software Engineer");
                matchedRoles.Add("Dotnet Developer");
            }
            else if (extractedSkills.Contains("Python") || extractedSkills.Contains("Machine Learning") || extractedSkills.Contains("Data Science"))
            {
                matchedRoles.Add("Data Scientist");
                matchedRoles.Add("Machine Learning Engineer");
            }
            else if (extractedSkills.Contains("React") || extractedSkills.Contains("TypeScript") || extractedSkills.Contains("HTML"))
            {
                matchedRoles.Add("Frontend Developer");
            }
            else
            {
                matchedRoles.Add("Software Developer");
            }
        }

        int experienceYears = ExtractExperienceYears(rawText);

        string education = "Bachelor's Degree";
        if (Regex.IsMatch(rawText, @"\b(Ph\.?D|Doctorate)\b", RegexOptions.IgnoreCase))
        {
            education = "Doctorate / PhD";
        }
        else if (Regex.IsMatch(rawText, @"\b(Master|M\.S|M\.Sc|M\.Tech|MBA)\b", RegexOptions.IgnoreCase))
        {
            education = "Master's Degree";
        }
        else if (Regex.IsMatch(rawText, @"\b(Bachelor|B\.S|B\.Sc|B\.Tech|B\.E)\b", RegexOptions.IgnoreCase))
        {
            education = "Bachelor's Degree";
        }

        bool isRemotePreferred = Regex.IsMatch(rawText, @"\b(Remote|Work from home|Hybrid)\b", RegexOptions.IgnoreCase);
        string preferredLocation = "Flexible / Remote";

        var locationMatch = Regex.Match(rawText, @"\b(London|Manchester|New York|San Francisco|Austin|Berlin|Toronto|Remote|UK|United Kingdom|USA)\b", RegexOptions.IgnoreCase);
        if (locationMatch.Success)
        {
            preferredLocation = locationMatch.Value;
        }

        return new CvProfile
        {
            UserId = userId,
            FileName = fileName,
            RawText = rawText,
            ExtractedSkillsJson = JsonSerializer.Serialize(extractedSkills.ToList()),
            TargetRolesJson = JsonSerializer.Serialize(matchedRoles.ToList()),
            ExperienceYears = experienceYears,
            EducationLevel = education,
            PreferredLocation = preferredLocation,
            IsRemotePreferred = isRemotePreferred,
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow
        };
    }

    private static int ExtractExperienceYears(string text)
    {
        var match = Regex.Match(text, @"(\d+)\+?\s*years?(?:\s+of)?\s+experience", RegexOptions.IgnoreCase);
        if (match.Success && int.TryParse(match.Groups[1].Value, out int yrs))
        {
            return Math.Min(yrs, 30);
        }

        var yearMatches = Regex.Matches(text, @"\b(20\d\d)\b");
        if (yearMatches.Count >= 2)
        {
            var years = yearMatches.Select(m => int.Parse(m.Value)).OrderBy(y => y).ToList();
            int span = years.Last() - years.First();
            if (span > 0 && span <= 25)
            {
                return span;
            }
        }

        return 0; // No experience specified -> 0 years (Graduate / Entry level)
    }
}

File: backend/app/services/cv_parsing_service.py
import io
import json
import logging
import re
from datetime import datetime
from pypdf import PdfReader
from docx import Document
from sqlalchemy.orm import Session
from fastapi import UploadFile, HTTPException, status
from app.models.models import CvProfile, User
from app.schemas.job_search import CvProfileDto
from app.services.gemini_cv_service import GeminiCvService

logger = logging.getLogger(__name__)


class CvParsingService:
    KNOWN_SKILLS = [
        "Python", "FastAPI", "Django", "Flask", "SQL", "PostgreSQL", "MySQL", "SQLite", "MongoDB", "Redis",
        "Docker", "Kubernetes", "AWS", "Azure", "GCP", "Git", "GitHub", "CI/CD", "REST API", "GraphQL",
        "Microservices", "React", "TypeScript", "JavaScript", "HTML", "CSS", "Tailwind CSS", "Next.js",
        "Node.js", "Express", "C#", ".NET", "Java", "Spring Boot", "Go", "Golang", "Rust", "C++",
        "Machine Learning", "Deep Learning", "PyTorch", "TensorFlow", "Scikit-learn", "Pandas", "NumPy",
        "Data Science", "Data Engineering", "Apache Spark", "Kafka", "Airflow", "NLP", "LLMs", "Computer Vision"
    ]

    def __init__(self, db: Session, gemini_service: GeminiCvService | None = None):
        self.db = db
        self.gemini_service = gemini_service or GeminiCvService()

    async def parse_and_save_cv(self, user_id: int, file: UploadFile) -> CvProfileDto:
        filename = file.filename or "uploaded_cv.pdf"
        content_bytes = await file.read()
        raw_text = self._extract_raw_text(filename, content_bytes)

        if not raw_text or len(raw_text.strip()) < 20:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Could not extract readable text from the uploaded CV file. Please ensure it is a valid PDF or DOCX file."
            )

        # Get user's gemini key if available
        user = self.db.query(User).filter(User.id == user_id).first()
        gemini_key = user.gemini_api_key if user else None

        extracted_skills = []
        target_roles = []
        experience_years = 0
        education_level = "Bachelor's Degree"
        preferred_location = "London, United Kingdom"
        is_remote = True

        # 1. Try Gemini AI extraction if configured
        gemini_data = await self.gemini_service.extract_cv_profile(raw_text, gemini_key)
        if gemini_data:
            extracted_skills = gemini_data.get("extractedSkills", [])
            target_roles = gemini_data.get("targetRoles", [])
            experience_years = gemini_data.get("experienceYears", 0)
            education_level = gemini_data.get("educationLevel", "Bachelor's Degree")
            preferred_location = gemini_data.get("preferredLocation", "London, United Kingdom")
            is_remote = gemini_data.get("isRemotePreferred", True)

        # 2. Rule-based extraction fallback/augmentation
        if not extracted_skills:
            extracted_skills = self._extract_skills_heuristic(raw_text)
        if not target_roles:
            target_roles = self._extract_target_roles_heuristic(raw_text)
        if experience_years == 0:
            experience_years = self._extract_experience_years_heuristic(raw_text)
        if not education_level or education_level == "Bachelor's Degree":
            education_level = self._extract_education_heuristic(raw_text)

        # Save or update CvProfile
        profile = self.db.query(CvProfile).filter(CvProfile.user_id == user_id).first()
        if not profile:
            profile = CvProfile(
                user_id=user_id,
                file_name=filename,
                raw_text=raw_text,
                extracted_skills=json.dumps(extracted_skills),
                target_roles=json.dumps(target_roles),
                experience_years=experience_years,
                education_level=education_level,
                preferred_location=preferred_location,
                is_remote_preferred=is_remote,
                created_at=datetime.utcnow(),
                updated_at=datetime.utcnow()
            )
            self.db.add(profile)
        else:
            profile.file_name = filename
            profile.raw_text = raw_text
            profile.extracted_skills = json.dumps(extracted_skills)
            profile.target_roles = json.dumps(target_roles)
            profile.experience_years = experience_years
            profile.education_level = education_level
            profile.preferred_location = preferred_location
            profile.is_remote_preferred = is_remote
            profile.updated_at = datetime.utcnow()

        self.db.commit()
        self.db.refresh(profile)

        return self.get_profile_dto(profile)

    def get_user_profile(self, user_id: int) -> CvProfileDto | None:
        profile = self.db.query(CvProfile).filter(CvProfile.user_id == user_id).first()
        return self.get_profile_dto(profile) if profile else None

    def get_profile_entity(self, user_id: int) -> CvProfile:
        profile = self.db.query(CvProfile).filter(CvProfile.user_id == user_id).first()
        if not profile:
            # Return default profile
            skills = ["Python", "FastAPI", "SQL", "PostgreSQL", "Docker", "AWS", "Git", "REST API", "React"]
            roles = ["Software Engineer", "Backend Developer", "Python Developer"]
            profile = CvProfile(
                user_id=user_id,
                file_name="Default_Profile.pdf",
                raw_text="Default Python Software Engineer Profile",
                extracted_skills=json.dumps(skills),
                target_roles=json.dumps(roles),
                experience_years=2,
                education_level="Bachelor's in Computer Science",
                preferred_location="London, United Kingdom",
                is_remote_preferred=True
            )
        return profile

    def get_profile_dto(self, profile: CvProfile) -> CvProfileDto:
        skills = json.loads(profile.extracted_skills) if profile.extracted_skills else []
        roles = json.loads(profile.target_roles) if profile.target_roles else []
        return CvProfileDto(
            id=profile.id,
            file_name=profile.file_name,
            extracted_skills=skills,
            target_roles=roles,
            experience_years=profile.experience_years,
            education_level=profile.education_level,
            preferred_location=profile.preferred_location,
            is_remote_preferred=profile.is_remote_preferred,
            uploaded_at=profile.updated_at or profile.created_at
        )

    def _extract_raw_text(self, filename: str, content: bytes) -> str:
        lower_name = filename.lower()
        if lower_name.endswith(".pdf"):
            try:
                reader = PdfReader(io.BytesIO(content))
                pages_text = [page.extract_text() or "" for page in reader.pages]
                return "\n".join(pages_text)
            except Exception as e:
                logger.error(f"Error reading PDF {filename}: {e}")
                return ""
        elif lower_name.endswith(".docx") or lower_name.endswith(".doc"):
            try:
                doc = Document(io.BytesIO(content))
                paragraphs = [p.text for p in doc.paragraphs if p.text]
                return "\n".join(paragraphs)
            except Exception as e:
                logger.error(f"Error reading DOCX {filename}: {e}")
                return ""
        else:
            try:
                return content.decode("utf-8", errors="ignore")
            except Exception:
                return ""

    def _extract_skills_heuristic(self, text: str) -> list[str]:
        found = []
        for skill in self.KNOWN_SKILLS:
            pattern = rf"\b{re.escape(skill)}\b"
            if re.search(pattern, text, re.IGNORECASE):
                found.append(skill)
        return found if found else ["Python", "FastAPI", "SQL", "Git"]

    def _extract_target_roles_heuristic(self, text: str) -> list[str]:
        candidates = [
            "Software Engineer", "Backend Developer", "Python Developer", "Full Stack Developer",
            "Data Scientist", "Machine Learning Engineer", "DevOps Engineer", "Cloud Engineer"
        ]
        found = []
        for role in candidates:
            if re.search(rf"\b{re.escape(role)}\b", text, re.IGNORECASE):
                found.append(role)
        return found if found else ["Software Engineer", "Backend Developer"]

    def _extract_experience_years_heuristic(self, text: str) -> int:
        match = re.search(r"(\d+)\+?\s*years?(?:\s+of)?\s+experience", text, re.IGNORECASE)
        if match:
            try:
                return int(match.group(1))
            except ValueError:
                pass
        return 2

    def _extract_education_heuristic(self, text: str) -> str:
        if re.search(r"\b(phd|doctorate)\b", text, re.IGNORECASE):
            return "PhD / Doctorate"
        if re.search(r"\b(master|msc|meng)\b", text, re.IGNORECASE):
            return "Master's Degree"
        if re.search(r"\b(bachelor|bsc|beng|b\.s\.|b\.tech)\b", text, re.IGNORECASE):
            return "Bachelor's Degree"
        return "Bachelor's Degree"
